044-cluster-wide-volume-usage-quota-management.md

```diff
@@ -0,0 +1,342 @@
+   1. throttle if available bytes less than threshold on any volume
+   2. throttle if available ratio less than threshold on any volume
+4. Introduce extension points in the quotas-plugin to support pluggable sources of quotas and throttle factors
+5. Introduce a [throttle factor fallback](#throttle-factor-fallback) and [throttle factor validity duration](#throttle-factor-validity-duration) in case we cannot retrieve the cluster state
+6. strimzi-kafka-operator becomes responsible for configuring the admin client connection properties of the plugin
+
+So every broker will make its own independent throttling decision based on knowledge of the volumes on all active broker nodes.
+The brokers should all operate on a similar view of the cluster state and make a deterministic decision. If a broker detects that
+any volume in the cluster is becoming too full it will throttle production of messages. The kafka quota API isn't rich 
+enough to do anything smarter about only throttling writes to the brokers running out of space, so we fence the whole cluster.
+
+### Caveat - KRaft Disk Usage
+
+This proposal will only help prevent running out-of-disk caused by topic data. It will not prevent disks being exhausted
+due to writes to the upcoming KRaft metadata log.
+
+Currently, when using separate controller-only nodes*, those nodes are not described by `Admin#describeCluster` and we cannot use 
+`Admin#describeLogDirs` against the controllers. So the disk usage of the volume could be invisible.
+
+When using controller+broker mode*, by default the metadata log is kept in the first `log.dirs` directory but could be 
+configured to a custom location, potentially on its own volume. If it was put on a separate volume it would not be
+described in the describeLogDirs response, nor contribute to the volume usage of an existing log dir.
+
+So in some cases with controller+broker mode it would afford some protection, as growth of the metadata log dir could
+cause throttling of topic writes (because they are on the same volume as another log dir).
+
+Even if we had reliable insight into the disk usage (by extending the admin apis for example), of the volume the metadata dir 
+resides on, we cannot take effective measures from the quota plugin. Potentially you could use an Authorization
+plugin instead and block metadata-generating operations.
+
+*tested with kafka 3.3.1
+### High level changes
+
+To better support alternative sources for managing quotas and how they interact with volume usage this proposal
+separates some existing concepts and separates their responsibilities. We envisage the architecture looking like
+this ![Component interactions](images/044-quota-plugin-interactions.png)
+
+#### Quota Source
+
+We propose adding a Quota Source concept to the plugin to provide an abstraction through which where we could provide
+external sources of quotas, like pull them from an external system.
+
+#### Throttle Factor
```

Comment:
I'm really confused how the throttle factor works now. In the `Limit Type Configuration` you don't distinguish anymore between soft and hard limit. So what is the factor based on now? Is it between the _limit- and full disk? Or how does it work now? I'm afraid it is not _implicit_ enough for me :-o.

Reply:
During normal operation (under the limit) the throttle factor would be 1.0, once the limit is reached it would become 0.0

 I guess we should state this somewhere, maybe like: `When that limit is exceeded we effectively stop messages being produced by setting the throttle factor to 0.0`

Reply:
So, the soft limit is essentially removed if I get it right and the traffic is either allowed or not allowed. Do I get it right?

I know we touch on this in the previous discussion here and do not have a problem with it. But, it might be good to:
* State it more clearly given we previously had this feature. You do not say anywhere that it is supported apart from the backwards compatibility parts. But it can be easily assumed from the current version.
* Perhaps update the example in `Throttle Factor Validity Duration` section? Because how would you get the throttle factor `0.6` if there is no soft limit?
* Also, it seems like the `Throttle Factor Fallback` is currently the only thing really using the throttle factor in an other way then just on/off switch? (no problem with that ... just trying to make sure if I understand it correctly or if I missed something)

Reply:
> soft limit is essentially removed if I get it right and the traffic is either allowed or not allowed

Correct.

> Also, it seems like the Throttle Factor Fallback is currently the only thing really using the throttle factor in an other way then just on/off switch

Correct.

Keeping it as a factor is now more about leaving it open to future extension to slow down writes if we found some reason to.

Will add a statement about the rejected soft limits.

Reply:
Pushed up a paragraph about how we are dropping the soft limit concept, and fixed the validity duration example.